cabal-validate/main/Main.hs

```diff
@@ -0,0 +1,953 @@
+                <> help "Run `hackage-tests` on complete Hackage data"
+            )
+            <|> flag
+              NoHackageTests
+              PartialHackageTests
+              ( long "partial-hackage-tests"
+                  <> help "Run `hackage-tests` on parts of Hackage data"
+              )
+        )
+
+-- | Parse a boolean switch with separate names for the true and false options.
+boolOption' :: Bool -> String -> String -> Mod FlagFields Bool -> Parser Bool
+boolOption' defaultValue trueName falseName modifiers =
+  flag' True (modifiers <> long trueName)
+    <|> flag defaultValue False (modifiers <> hidden <> long falseName)
+
+-- | Parse a boolean switch with a `--no-*` flag for setting the option to false.
+boolOption :: Bool -> String -> Mod FlagFields Bool -> Parser Bool
+boolOption defaultValue trueName =
+  boolOption' defaultValue trueName ("no-" <> trueName)
+
+fullOptsParser :: ParserInfo Opts
+fullOptsParser =
+  info
+    (optsParser <**> helper)
+    ( fullDesc
+        <> progDesc "Test suite runner for `Cabal` and `cabal-install` developers"
+    )
+
+data HackageTests
+  = CompleteHackageTests
+  | PartialHackageTests
+  | NoHackageTests
+  deriving (Show)
+
+data Step
+  = PrintConfig
+  | PrintToolVersions
+  | Build
+  | Doctest
+  | LibTests
+  | LibSuite
+  | LibSuiteExtras
+  | CliTests
+  | CliSuite
+  | SolverBenchmarksTests
+  | SolverBenchmarksRun
+  | TimeSummary
+  deriving (Eq, Enum, Bounded, Show)
+
+displayStep :: Step -> Text
+displayStep step =
+  case step of
+    PrintConfig -> "print-config"
+    PrintToolVersions -> "print-tool-versions"
+    Build -> "build"
+    Doctest -> "doctest"
+    LibTests -> "lib-tests"
+    LibSuite -> "lib-suite"
+    LibSuiteExtras -> "lib-suite-extras"
+    CliTests -> "cli-tests"
+    CliSuite -> "cli-suite"
+    SolverBenchmarksTests -> "solver-benchmarks-tests"
+    SolverBenchmarksRun -> "solver-benchmarks-run"
+    TimeSummary -> "time-summary"
+
+nameToStep :: Map Text Step
+nameToStep =
+  Map.fromList
+    [ (displayStep step, step)
+    | step <- [minBound .. maxBound]
+    ]
+
+parseStep :: Text -> Maybe Step
+parseStep step = Map.lookup step nameToStep
+
+runStep :: ResolvedOpts -> Step -> IO ()
+runStep opts step = do
+  let title = displayStep step
+  printHeader title
+  let action = case step of
+        PrintConfig -> printConfig opts
+        PrintToolVersions -> printToolVersions opts
+        Build -> build opts
+        Doctest -> doctest opts
+        LibTests -> libTests opts
+        LibSuite -> libSuite opts
+        LibSuiteExtras -> libSuiteExtras opts
+        CliSuite -> cliSuite opts
+        CliTests -> cliTests opts
+        SolverBenchmarksTests -> solverBenchmarksTests opts
+        SolverBenchmarksRun -> solverBenchmarksRun opts
+        TimeSummary -> timeSummary opts
+  withTiming opts title action
+  tPutStrLn ""
+
+getTerminalWidth :: IO Int
+getTerminalWidth = maybe 80 width <$> size @Int
```

Comment:
Could `terminal-size` to be imported qualified? Otherwise it's hard to guess what kind of `size` it is.